### PR TITLE
Remove include dependency from 'gen/enums.h' inside src/app

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
non-generated source code depends on generated code, which prevents us from cleanly building a stand-alone chip library as well as from neatly integrating GN (have to add source sets instead of stand alone libraries for layering)

 #### Summary of Changes
I replace the include for generated types with a copy&pasted version of enums that contains what seem to be hard-coded enum values that are always available in gen.

Assumption is that these enums always need to be available (otherwise code would not compile) and that they always have the same value (seems correct so far, but have to ensure later).

Working on #4397 and #1630

### CODE REVIEW

**Comment by andy31415** [comment]
I believe the generation of these enums should also be pulled out from ZAP ... needs to be done separately and I have never patched zap so far - would appreciate some help there.

**Review by mspang — APPROVED** [review]

**Review by rwalker-apple — APPROVED** [review]

**Comment by rwalker-apple** [comment]
@saurabhst @hawk248 @jelderton @BroderickCarlin ?

**Comment by vivien-apple** [comment]
> I believe the generation of these enums should also be pulled out from ZAP ... needs to be done separately and I have never patched zap so far - would appreciate some help there.

It seems OK for the `direction` and `status`since I guess they are not going to change much and are global values. I'm more reluctant on the `TariffType` thing. Mind removing `EmberAfDeviceManagementInfo` struct from `src/app/util/af-types.h` instead ? It is not used affect and I think this is the reason why you need to move `EmberAfTariffType`.

Now in order to fix the ZAP generation, you should normally just remove those lines from the Silabs definitions of ZCL:
 * https://github.com/project-chip/zap/blob/f462cc913114a148ebb52b83110daf61e9945594/zcl-builtin/silabs/types.xml#L213
 *  https://github.com/project-chip/zap/blob/master/zcl-builtin/silabs/types.xml#L615
 * https://github.com/project-chip/zap/blob/master/zcl-builtin/silabs/ami.xml#L367 (again I suggest to not change this one. It sounds random).

Now, I guess the issue is the fact, that so far I've tried to avoid to duplicate the `zcl-builtin` folder of ZAP into the CHIP source tree in order to benefit from the possible upstream changes.
Here are the possible solutions on top of my mind to properly remove those lines:
 * Do the change upstream. I believe `Silabs` may want to keep that to be in sync with some possible other consumers of `ZAP`
 *  Copy-Paste the `zcl-builtin/silabs` folder from ZAP and start to do our own version (but that seems overkill at the moment)
 *  Copy-Paste the `xcl-builtin/silabs/types.xml` file into `src/app/zap-templates/zcl/`and update `src/app/zap-templates/zcl/zcl.json` to use this version instead of the one from Silabs.
 *  Do a dirty trick inside the templates to avoid to generate the enums for those.
 * Split the upstream `zcl-builtin/silabs/types.xml` into 2 files, one with the enums we are interested in, and one with the enum we don't want. That would let us select the right file to pick in our local `zcl.json` file <- **preferred option**

**Comment by andy31415** [comment]
@vivien-apple - I removed the device info and tariff type.

For long term solution, I believe we should have a separate 'generated files for chilLib' that allows to build chip-files and then the application-specific side which users can use to generate further things (like custom clusters).

**Comment by vivien-apple** [comment]
> @vivien-apple - I removed the device info and tariff type.
> 
> For long term solution, I believe we should have a separate 'generated files for chipLib' that allows to build chip-files and then the application-specific side which users can use to generate further things (like custom clusters).

From what I know, all of the `generated files for chipLib` that we have at the moment in the tree are related to the `client` side of clusters.
I believe they exists at the moment because none of our client applications (`chip-tool`, python device controller, iOS app, Android app) have their own `.zap` file.  Instead they rely on generated code living inside `src/app`.

This is not reliable since it won't work for custom clusters, custom commands, custom attributes, etc...
 
My hope is that we can get rid of `src/app/encoder.cpp`,  `src/controller/CHIPClusters.*`,  `src/darwin/Framework/CHIP/CHIPClustersObjc.*` from `src/app` and instead move them into a client application `gen/` folder. Similar to the `server` side of things.

In short, I'm not convinced any generated code should be part of `chip` directly.

I should have a version of `chip-tool` that uses its own `.zap` file soon (maybe the end of the week). It is mostly working, still needs to generate a few things related to callbacks, and maybe it will make things more understandable.

**Comment by vivien-apple** [comment]
Also we still need to fix the `ZAP` side of things if we want to not have to manually fix the files all the time.
@jepenven-silabs Does splitting `types.xml` into 2 files sounds reasonable to you ?

one that contains:
 * https://github.com/project-chip/zap/blob/f462cc913114a148ebb52b83110daf61e9945594/zcl-builtin/silabs/types.xml#L213
 *  https://github.com/project-chip/zap/blob/master/zcl-builtin/silabs/types.xml#L615

And the other that contains the rest of `types.xml`

It would let `chip` internals have less dependencies on the generated code by internalising a very small subsets of enums by only picking the interesting file in our `zcl.json` config, while I believe it won't change anything on the generator side.

**Comment by jepenven-silabs** [comment]
@vivien-apple Splitting the xml in two seems to be like a good idea. A small note on `/zcl-builtin/silabs` We SHOULD make a copy of our ZCL configuration within the CHIP repo. Point is that the XML present in this folder are uniquely there for unit-test purposes and can be change by silabs at any time to fit our needs. We used them as is for now without any problem but trusting that they will stay there and remain unchanged is a bad assumption. Moreover with the addition of a few new cluster in the CHIP specs it would be easier to track changes if all of the XML that chip uses are stored in one place.

**Comment by andy31415** [comment]
this got merged as part of the followup (#4409), closing it.

Still need the followups though - I will look into zap but there will be some rampup time.
